salt,dex: Allow to disable Dex deployment and configure another IDP

alexandre-allard opened the pull request:
**Component**: salt, dex

**Context**:
We want to be able to not deploy Dex and rather rely on an external IDP.

**Summary**:

To disable Dex deployment at installation,
you must add the following to the boostrap config:

```
addons:
  dex:
    enabled: false
```

**Acceptance criteria**: 
We can use another IDP than Dex (e.g. Keycloak) and everything work fine.

alexandre-allard committed:
salt,dex: Allow to disable Dex deployment

To disable Dex deployment at installation,
you must add the following to the boostrap config:

```
addons:
  dex:
    enabled: false
```

bert-e commented:
# Hello alexandre-allard-scality,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @chengyanjin

* @lucieleonard

* @jbwatenbergscality

* @alexandre-allard-scality

* @gdemonet

* @teddyandrieux

* @nicolast

* @thomasdanan

alexandre-allard committed:
salt: Dynamic rendering of Grafana INI config

Extract Grafana INI configuration and put it
in a CSC ConfigMap to allow dynamic configuration.

Re-render the chart with the following command:
```
./charts/render.py prometheus-operator \
  charts/kube-prometheus-stack.yaml \
  charts/kube-prometheus-stack/ \
  --namespace metalk8s-monitoring \
  --service-config grafana \
  metalk8s-grafana-config \
  metalk8s/addons/prometheus-operator/config/grafana.yaml.j2 \
  metalk8s-monitoring \
  --service-config prometheus \
  metalk8s-prometheus-config \
  metalk8s/addons/prometheus-operator/config/prometheus.yaml \
  metalk8s-monitoring \
  --service-config alertmanager \
  metalk8s-alertmanager-config \
  metalk8s/addons/prometheus-operator/config/alertmanager.yaml \
  metalk8s-monitoring \
  --drop-prometheus-rules charts/drop-prometheus-rules.yaml \
  --patch 'PrometheusRule,metalk8s-monitoring,prometheus-operator-kubernetes-system-kubelet,spec:groups:0:rules:1:for,"5m"' \
  --remove-manifest ConfigMap prometheus-operator-grafana \
  > salt/metalk8s/addons/prometheus-operator/deployed/chart.sls
```

alexandre-allard committed:
salt: Allow to configure OIDC for api-server

gdemonet commented on salt/tests/unit/modules/files/test_metalk8s_service_configuration.yaml:
```diff
@@ -64,6 +64,24 @@ get_service_config:
           containers:
             - name: my_new_container
 
+  #. Success: retrieved ConfigMap manifest is empty
```
That's quite the functional change, any idea about the impact? Are we properly handling this case in formulas?

TeddyAndrieux replied:
I think, we do, since we never catch the exception today so we always expect the ConfigMap to exists

gdemonet commented on salt/metalk8s/addons/prometheus-operator/config/grafana.yaml.j2:
```diff
@@ -0,0 +1,45 @@
+      log:
+        mode: console
+      grafana_net:
+        url: https://grafana.net
```
Let's drop this, we really don't need it IMO: https://grafana.com/docs/grafana/latest/administration/configuration/#grafana_net (they don't even explain what it's for)

gdemonet commented on salt/metalk8s/addons/prometheus-operator/config/grafana.yaml.j2:
```diff
@@ -0,0 +1,45 @@
+    grafana.ini:
+      analytics:
+        check_for_updates: true
+        reporting_enabled: falseœ
```
That last character `œ` shouldn't be here

alexandre-allard replied:
Ha! It was just to see if you really review the PR!

gdemonet commented on salt/metalk8s/addons/prometheus-operator/config/grafana.yaml.j2:
```diff
@@ -0,0 +1,45 @@
+  config:
+    grafana.ini:
+      analytics:
+        check_for_updates: true
```
Why is it now `true`?

alexandre-allard replied:
That's a good question, someone has hacked my github account and changed it I presume...

TeddyAndrieux commented on salt/metalk8s/addons/dex/deployed/init.sls:
```diff
@@ -9,6 +9,8 @@
 # * chart                  -> charts used to deploy Dex
 # * clusterrolebinding     -> binds dex static user to cluster admin
 
+{%- if pillar.addons.dex.enabled %}
```
Nitpick: To me, it could be done in `metalk8s.deployed` state directly so that you can manually deploy Dex running the state but it's not deployed by bootstrap, upgrade, ...

alexandre-allard replied:
Well, it's what I did in the beginning, but the goal was to be sure we can't force or mistakenly deploy it.
For example, in the MetalK8s cluster and services configurations documentation, we explicitly say to run `metalk8s.addons.dex.deployed` after a change to its related CSC ConfigMap.
And we do it in the ARTESCA installer for both Keycloak and MetalK8s configuration//deployment.

TeddyAndrieux replied:
Right ok make sense

TeddyAndrieux commented on salt/metalk8s/addons/ui/deployed/init.sls:
```diff
@@ -1,6 +1,15 @@
+{%- if pillar.addons.dex.enabled %}
```
ditto, it could be in `metalk8s.deployed` state

TeddyAndrieux commented on salt/metalk8s/addons/prometheus-operator/deployed/grafana-ini-configmap.sls:
```diff
@@ -0,0 +1,30 @@
+          namespace: metalk8s-monitoring
+        data:
+          grafana.ini: |-
+{%- for key, value in grafana.spec.config["grafana.ini"].items() %}
+            [{{ key }}]
+  {%- for element, element_value in value.items() %}
+            {{ element }} = {{ element_value }}
+  {%- endfor %}
+{%- endfor %}
```
Uh :confused: this is really needed ? A `| yaml(False) | indent(12)` does not work ?

alexandre-allard replied:
It's not YAML, but INI configuration file.

alexandre-allard replied:
But we could probably add a Jinja filter to format INI configurations.

TeddyAndrieux replied:
Oh you're right :cry: 
Ok no I think it's good like that then

TeddyAndrieux commented on salt/tests/unit/formulas/data/base_pillar.yaml:
```diff
@@ -187,3 +187,4 @@ kubernetes:
 addons:
   dex:
     enabled: True
+oidc: {}
```
Maybe worth adding a unit test with an OIDC in pillar content (just to ensure that we do not break the salt state rendering)

alexandre-allard replied:
done

TeddyAndrieux commented on buildchain/buildchain/salt_tree.py:
```diff
@@ -351,6 +351,7 @@ def task(self) -> types.TaskDict:
         "alertmanager-configuration-secret.sls",
     ),
     Path("salt/metalk8s/addons/prometheus-operator/deployed/chart.sls"),
+    Path("salt/metalk8s/addons/prometheus-operator/deployed/chart-without-grafana.sls"),
```
TBH: I'm not sure we need to do this... it means yet another state to maintain (and test, ...),

Since the grafana config can be managed by the user himself, we shouldn't care about it and always deploy grafana. (what happened if we deploy grafana without `auth` section ?)

And maybe it should be the same for the metalk8s-ui depending on how the pod behaves when you do not have OIDC (crash loop or just the web UI that show an error ?)

alexandre-allard replied:
Everything works if we still deploy Grafana and MetalK8s UI, I only did this because we said we don't want to deploy them without an OIDC in the meeting we did.

TeddyAndrieux replied:
Ok, yes I see, but now to me, we shouldn't deploy Dex as it will not be used, but deploy Grafana and MetalK8s UI even if it's still not usable for the moment... it's ok and easier to maintain

Maybe it should be re-discussed a bit with @scality/metalk8s 

I just do not see a real value of not deploying them. Especially for Grafana where we will need to "maintain" a new salt state (and ensure this one works properly)

gdemonet replied:
Good point @TeddyAndrieux, we could just deploy and expect things not to work (on client-side)... For the UX aspect, however, I'd like to review how it looks like when we try to load these UIs in the browser.

TeddyAndrieux replied:
Already discussed offline on Slack and fixed in this PR + a ticket has been opened for MetalK8s UI

TeddyAndrieux reviewed (commented):
Overall LGTM but I think we will need to add some E2E tests maybe just a really simple one for the moment
Like just a simple install with no OIDC for the moment, we put it in post-merge so that we can enrich it in the future with a real scenario (deploy with no OIDC, then set up another OIDC then re-configure the product to use this new OIDC)

gdemonet commented on salt/metalk8s/addons/prometheus-operator/config/grafana.yaml.j2:
```diff
@@ -0,0 +1,45 @@
+        enabled: true
+        role_attribute_path: contains(`{{ dex.spec.config.staticPasswords | map(attribute='email') | list | tojson }}`, email) && 'Admin'
+        scopes: openid profile email groups
+        tls_skip_verify_insecure: true
```
Oh really? :disappointed: (not something you need to take care of, but we'll have to open a ticket)

TeddyAndrieux replied:
Done

gdemonet commented on salt/metalk8s/kubernetes/apiserver/installed.sls:
```diff
@@ -96,15 +109,15 @@ Create kube-apiserver Pod manifest:
-          - --oidc-groups-claim=groups
-          {%- endif %}
+          {%- if oidc_config %}
+          - --oidc-issuer-url={{ oidc_config.issuerURL }}
+          - --oidc-client-id={{ oidc_config.clientID }}
+          - --oidc-ca-file={{ oidc_config.CAFile }}
+          - --oidc-username-claim={{ oidc_config.usernameClaim }}
+          - --oidc-groups-claim={{ oidc_config.groupsClaim }}
```
Since these are k-a settings only, wouldn't it make more sense to define in `pillar.kubernetes` instead?

alexandre-allard replied:
done

TeddyAndrieux committed:
eve: Add post-merge e2e test for install without IDP

We add a new stage in post-merge to test installation without Dex, note
that this stage would likely be updated in the future in order to deploy
another IDP and use this new IDP for Grafana and MetalK8s UI

TeddyAndrieux committed:
tests: Add "authentication" tag on test that rely on Dex

Since we can deploy MetalK8s without Dex we need to be able to
skip test that rely on Dex

TeddyAndrieux committed:
changelog: Add entry for installation without Dex

gdemonet commented on salt/metalk8s/addons/prometheus-operator/config/grafana.yaml.j2:
```diff
@@ -0,0 +1,44 @@
+      server:
+        root_url: "{{ control_plane_ingress_endpoint }}/grafana"
+      auth:
+        disable_login_form: true
```
That's worth documenting in the changelog, no?

TeddyAndrieux replied:
To me no, it does not change anything compared to the current behavior since anyway when we used dex `oauth_auto_login` was set to True so we never reached the Grafana login form

gdemonet replied:
Oh good point :+1:

gdemonet commented on docs/installation/bootstrap.rst:
```diff
@@ -207,6 +208,20 @@ defaults kubernetes configuration.
   configure the corresponding entries in the
   ``kubernetes.apiServer.featureGates`` mapping.
 
+  If ``dex`` is enabled, it will be used as ``oidc`` for ``kube-apiserver``
+  but you can use a `specific OpenID for kube-apiserver`_, to do so:
```
Maybe mention this in the paragraph above about disabling dex, no?

TeddyAndrieux replied:
Hmm right I will mention apiserver as well

TeddyAndrieux replied:
Btw I'm not sure it really makes sense to put this in the dex section,
Because if you disable Dex there is no impact on apiserver itself, except that it will not use Dex but it's obvious. To me what we want to reflect is that oidc for apiserver is configurable and default to dex when it's deployed. 
WDYT ?

gdemonet replied:
Ah right, we only have Grafana / MetalK8s UI if we use Dex, today at least... So indeed, no need to add anything to the addons paragraph, until we manage these.

gdemonet commented on CHANGELOG.md:
```diff
@@ -1,6 +1,12 @@
+### Additions
+
+- Add ability to deploy MetalK8s without Dex and possibility to configure
+  our own IDP for kube-apiserver, Grafana and MetalK8s UI
```
```suggestion
  your own IDP for kube-apiserver, Grafana and MetalK8s UI
```

TeddyAndrieux reviewed (approved):
~/bypass_author_approval~ (it does not work in review comment apparently)

TeddyAndrieux commented:
/bypass_author_approval

bert-e commented:
# Build failed



The build for [commit](https://github.com/scality/metalk8s/commit/48d37343277ed384c4482a1012dc41ea7156973f) did not succeed in branch improvement/no-idp.




*The following options are set:* **bypass_author_approval**

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/123.0`



The following branches will **NOT** be impacted:

* `development/2.0`
* `development/2.1`
* `development/2.10`
* `development/2.11`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`
* `development/2.6`
* `development/2.7`
* `development/2.8`
* `development/2.9`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **bypass_author_approval**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/123.0`



The following branches have **NOT** changed:

* `development/2.0`
* `development/2.1`
* `development/2.10`
* `development/2.11`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`
* `development/2.6`
* `development/2.7`
* `development/2.8`
* `development/2.9`



Please check the status of the associated issue None.

Goodbye alexandre-allard.
